src/endpoint/s3/s3_utils.js

```diff
@@ -616,6 +613,16 @@ function parse_decimal_int(str) {
     return parsed;
 }
 
+/**
+ * isEncoded consumes returns true if string uri is URIEncoded
+ * @param {string} uri
+ * @returns {boolean}
+ */
+function isEncoded(uri) {
+    uri = uri || '';
```

Comment:
Use default in function args, renamed, and added also encode_uri_unless_already_encoded() which deserves its own function

```suggestion
function encode_uri_unless_already_encoded(uri = '') {
    return is_uri_already_encoded(uri) ? uri : encodeURIComponent(uri);
}

function is_uri_already_encoded(uri = '') {
```